aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,356 @@
+      typename = std::enable_if_t<
+          std::is_constructible<unboxed_type, UnboxedConstructorArgs...>::value>>
+  ITensorList(UnboxedConstructorArgs&&... args)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = unboxed_type(std::forward<UnboxedConstructorArgs>(args)...);
+  }
+
+  size_t size() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.size(); });
+  }
+
+  bool empty() const {
+    return size() == 0;
+  }
+
+  iterator begin() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.begin(); });
+  }
+
+  iterator end() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.end(); });
+  }
+
+  detail::ITensorListConstRef operator[](size_t i) const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_[i]; });
+  }
+
+#define DEFINE_CHECK(TAG, ...)          \
+  bool is##TAG() const {                \
+    return tag_ == ITensorListTag::TAG; \
+  }
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_CHECK);
+#undef DEFINE_CHECK
+
+  bool isNone() const {
+    return tag_ == ITensorListTag::None;
+  }
+
+#define DEFINE_CASTING(TAG, ...)                                        \
+  const typename TORCH_ITENSORLIST_IMPL(TAG)::list_type& to##TAG() const { \
+    TORCH_INTERNAL_ASSERT(is##TAG());                                   \
+    return TORCH_ITENSORLIST_IMPL(TAG)::unwrap(*this);                  \
+  }
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_CASTING);
+#undef DEFINE_CASTING
+
+ private:
+  Payload payload_;
+  ITensorListTag tag_;
+};
+} // namespace c10
+
+inline
+const TORCH_ITENSORLIST_IMPL(Unboxed)::list_type&
+TORCH_ITENSORLIST_IMPL(Unboxed)::unwrap(
+    const c10::ITensorList& ilist
+) {
+  return ilist.payload_.unboxed;
+}
+
+inline
+TORCH_ITENSORLIST_IMPL(Unboxed)::list_type::const_iterator&
+TORCH_ITENSORLIST_IMPL(Unboxed)::unwrap(
+    c10::ITensorListIterator& it
+) {
+  return it.payload_.unboxed_iterator;
+}
+
+inline
+const TORCH_ITENSORLIST_IMPL(Unboxed)::list_type::const_iterator&
+TORCH_ITENSORLIST_IMPL(Unboxed)::unwrap(
+    const c10::ITensorListIterator& it
+) {
+  return it.payload_.unboxed_iterator;
+}
+
+inline
+c10::detail::ITensorListConstRef
+TORCH_ITENSORLIST_IMPL(Unboxed)::iterator_get(
+    const list_type::const_iterator& it
+) {
+  return *it;
+}
+
+inline
+const TORCH_ITENSORLIST_IMPL(Boxed)::list_type&
+TORCH_ITENSORLIST_IMPL(Boxed)::unwrap(
+    const c10::ITensorList& ilist
+) {
+  return *ilist.payload_.boxed;
+}
+
+inline
+TORCH_ITENSORLIST_IMPL(Boxed)::list_type::const_iterator&
+TORCH_ITENSORLIST_IMPL(Boxed)::unwrap(
+    c10::ITensorListIterator& it
+) {
+  return it.payload_.boxed_iterator;
+}
+
+inline
+const TORCH_ITENSORLIST_IMPL(Boxed)::list_type::const_iterator&
+TORCH_ITENSORLIST_IMPL(Boxed)::unwrap(
+    const c10::ITensorListIterator& it
+) {
+  return it.payload_.boxed_iterator;
+}
+
+inline
+c10::detail::ITensorListConstRef
+TORCH_ITENSORLIST_IMPL(Boxed)::iterator_get(
+    const list_type::const_iterator& it
+) {
+  return (*it).get().toTensor();
+}
```

Comment:
The interface here is pretty wide. One thing I'll be looking for here is whether or not any of the API provided here encourage 'inefficient' access patterns; ideally, we want users to iterate through the list once taking out const references each time, and then that is it; repeatedly shoving their fingers in is bad.